### PR TITLE
[SYSTEMML-540] Support sparse GPU conv2d as well as fix memory estimation of convolution operations

### PR DESCRIPTION
**Pull request by niketanpansare**
Design doc: Memory estimation of GPU operators

1. Since not all operator are supported on GPU, isGPUEnabled indicates whether an operation  is enabled for GPU. This method doesnot take into account any memory estimates.
2. To simplify memory estimation logic, the methods computeOutputMemEstimate and computeIntermediateMemEstimate should return maximum of memory required for GPU and CP operators. 
3. Additionally, these methods are guarded so that when -gpu flag is not provided, additional memory overhead due to GPU are ignored. For example: sparse-to-dense conversion on GPU. 
4. (WIP) Every GPU operators should respect the memory returned by computeIntermediateMemEstimate (and computeOutputMemEstimate - see below point).
5. (WIP) Every GPU operator should create output in the same format as the corresponding CP operator. That is,  computeOutputMemEstimate are consistent across both CP and GPU in terms of worst-case.
6. The drawback of using maximum memory (mem = Math.max(mem_gpu, mem_gpu)) are:
- GPU operator is not selected when mem_gpu < total memory available on GPU < mem
 - CP operator is not selected (i.e. distributed operator compiled) when mem_cpu < driver memory budget < mem

@nakul02

### CODE REVIEW

**Comment by akchinSTC** [comment]
Refer to this link for build results (access rights to CI server needed): 
https://sparktc.ibmcloud.com/jenkins/job/SystemML-PullRequestBuilder/2034/<h2>Failed Tests: <span class='status-failure'>70</span></h2><h3><a name='SystemML-PullRequestBuilder/org.apache.systemml:systemml' /><a href='https://sparktc.ibmcloud.com/jenkins/job/SystemML-PullRequestBuilder/2034/org.apache.systemml$systemml/testReport'>SystemML-PullRequestBuilder/org.apache.systemml:systemml</a>: <span class='status-failure'>70</span></h3><ul><li><a href='https://sparktc.ibmcloud.com/jenkins/job/SystemML-PullRequestBuilder/2034/org.apache.systemml$systemml/testReport/org.apache.sysml.test.integration.functions.misc/IPAConstantFoldingScalarVariablePropagationTest/testConstantFoldingScalarPropagation2IPASecondChance/'><strong>org.apache.sysml.test.integration.functions.misc.IPAConstantFoldingScalarVariablePropagationTest.testConstantFoldingScalarPropagation2IPASecondChance</strong></a></li><li><a href='https://sparktc.ibmcloud.com/jenkins/job/SystemML-PullRequestBuilder/2034/org.apache.systemml$systemml/testReport/org.apache.sysml.test.integration.functions.misc/IPAConstantFoldingScalarVariablePropagationTest/testConstantFoldingScalarPropagation1IPASecondChance/'><strong>org.apache.sysml.test.integration.functions.misc.IPAConstantFoldingScalarVariablePropagationTest.testConstantFoldingScalarPropagation1IPASecondChance</strong></a></li><li><a href='https://sparktc.ibmcloud.com/jenkins/job/SystemML-PullRequestBuilder/2034/org.apache.systemml$systemml/testReport/org.apache.sysml.test.integration.functions.misc/IPAConstantFoldingScalarVariablePropagationTest/testConstantFoldingScalarPropagation2NoIPASecondChance/'><strong>org.apache.sysml.test.integration.functions.misc.IPAConstantFoldingScalarVariablePropagationTest.testConstantFoldingScalarPropagation2NoIPASecondChance</strong></a></li><li><a href='https://sparktc.ibmcloud.com/jenkins/job/SystemML-PullRequestBuilder/2034/org.apache.systemml$systemml/testReport/org.apache.sysml.test.integration.functions.misc/IPAConstantFoldingScalarVariablePropagationTest/testConstantFoldingScalarPropagation1NoIPASecondChance/'><strong>org.apache.sysml.test.integration.functions.misc.IPAConstantFoldingScalarVariablePropagationTest.testConstantFoldingScalarPropagation1NoIPASecondChance</strong></a></li><li><a href='https://sparktc.ibmcloud.com/jenkins/job/SystemML-PullRequestBuilder/2034/org.apache.systemml$systemml/testReport/org.apache.sysml.test.integration.functions.tensor/Conv2DBackwardDataTest/testConv2DBwdDataSparse1/'><strong>org.apache.sysml.test.integration.functions.tensor.Conv2DBackwardDataTest.testConv2DBwdDataSparse1</strong></a></li><li><a href='https://sparktc.ibmcloud.com/jenkins/job/SystemML-PullRequestBuilder/2034/org.apache.systemml$systemml/testReport/org.apache.sysml.test.integration.functions.tensor/Conv2DBackwardDataTest/testConv2DBwdDataSparse2/'><strong>org.apache.sysml.test.integration.functions.tensor.Conv2DBackwardDataTest.testConv2DBwdDataSparse2</strong></a></li><li><a href='https://sparktc.ibmcloud.com/jenkins/job/SystemML-PullRequestBuilder/2034/org.apache.systemml$systemml/testReport/org.apache.sysml.test.integration.functions.tensor/Conv2DBackwardDataTest/testConv2DBwdDataSparse3/'><strong>org.apache.sysml.test.integration.functions.tensor.Conv2DBackwardDataTest.testConv2DBwdDataSparse3</strong></a></li><li><a href='https://sparktc.ibmcloud.com/jenkins/job/SystemML-PullRequestBuilder/2034/org.apache.systemml$systemml/testReport/org.apache.sysml.test.integration.functions.tensor/Conv2DBackwardDataTest/testConv2DBwdDataSparse4/'><strong>org.apache.sysml.test.integration.functions.tensor.Conv2DBackwardDataTest.testConv2DBwdDataSparse4</strong></a></li><li><a href='https://sparktc.ibmcloud.com/jenkins/job/SystemML-PullRequestBuilder/2034/org.apache.systemml$systemml/testReport/org.apache.sysml.test.integration.functions.tensor/Conv2DBackwardDataTest/testConv2DBwdDataSparse5/'><strong>org.apache.sysml.test.integration.functions.tensor.Conv2DBackwardDataTest.testConv2DBwdDataSparse5</strong></a></li><li><a href='https://sparktc.ibmcloud.com/jenkins/job/SystemML-PullRequestBuilder/2034/org.apache.systemml$systemml/testReport/org.apache.sysml.test.integration.functions.tensor/Conv2DBackwardDataTest/testConv2DBwdDataSparse6/'><strong>org.apache.sysml.test.integration.functions.tensor.Conv2DBackwardDataTest.testConv2DBwdDataSparse6</strong></a></li><li><a href='https://sparktc.ibmcloud.com/jenkins/job/SystemML-PullRequestBuilder/2034/org.apache.systemml$systemml/testReport/org.apache.sysml.test.integration.functions.tensor/Conv2DBackwardDataTest/testConv2DBwdDataSparse7/'><strong>org.apache.sysml.test.integration.functions.tensor.Conv2DBackwardDataTest.testConv2DBwdDataSparse7</strong></a></li><li><a href='https://sparktc.ibmcloud.com/jenkins/job/SystemML-PullRequestBuilder/2034/org.apache.systemml$systemml/testReport/org.apache.sysml.test.integration.functions.tensor/Conv2DBackwardDataTest/testConv2DDense2/'><strong>org.apache.sysml.test.integration.functions.tensor.Conv2DBackwardDataTest.testConv2DDense2</strong></a></li><li><a href='https://sparktc.ibmcloud.com/jenkins/job/SystemML-PullRequestBuilder/2034/org.apache.systemml$systemml/testReport/org.apache.sysml.test.integration.functions.tensor/Conv2DBackwardDataTest/testConv2DDense3/'><strong>org.apache.sysml.test.integration.functions.tensor.Conv2DBackwardDataTest.testConv2DDense3</strong></a></li><li><a href='https://sparktc.ibmcloud.com/jenkins/job/SystemML-PullRequestBuilder/2034/org.apache.systemml$systemml/testReport/org.apache.sysml.test.integration.functions.tensor/Conv2DBackwardDataTest/testConv2DBwdDataDense1/'><strong>org.apache.sysml.test.integration.functions.tensor.Conv2DBackwardDataTest.testConv2DBwdDataDense1</strong></a></li><li><a href='https://sparktc.ibmcloud.com/jenkins/job/SystemML-PullRequestBuilder/2034/org.apache.systemml$systemml/testReport/org.apache.sysml.test.integration.functions.tensor/Conv2DBackwardDataTest/testConv2DBwdDataDense4/'><strong>org.apache.sysml.test.integration.functions.tensor.Conv2DBackwardDataTest.testConv2DBwdDataDense4</strong></a></li><li><a href='https://sparktc.ibmcloud.com/jenkins/job/SystemML-PullRequestBuilder/2034/org.apache.systemml$systemml/testReport/org.apache.sysml.test.integration.functions.tensor/Conv2DBackwardTest/testConv2DBackwardFilterSparse1/'><strong>org.apache.sysml.test.integration.functions.tensor.Conv2DBackwardTest.testConv2DBackwardFilterSparse1</strong></a></li><li><a href='https://sparktc.ibmcloud.com/jenkins/job/SystemML-PullRequestBuilder/2034/org.apache.systemml$systemml/testReport/org.apache.sysml.test.integration.functions.tensor/Conv2DBackwardTest/testConv2DBackwardFilterSparse2/'><strong>org.apache.sysml.test.integration.functions.tensor.Conv2DBackwardTest.testConv2DBackwardFilterSparse2</strong></a></li><li><a href='https://sparktc.ibmcloud.com/jenkins/job/SystemML-PullRequestBuilder/2034/org.apache.systemml$systemml/testReport/org.apache.sysml.test.integration.functions.tensor/Conv2DBackwardTest/testConv2DBackwardFilterSparse3/'><strong>org.apache.sysml.test.integration.functions.tensor.Conv2DBackwardTest.testConv2DBackwardFilterSparse3</strong></a></li><li><a href='https://sparktc.ibmcloud.com/jenkins/job/SystemML-PullRequestBuilder/2034/org.apache.systemml$systemml/testReport/org.apache.sysml.test.integration.functions.tensor/Conv2DBackwardTest/testConv2DBackwardFilterSparse4/'><strong>org.apache.sysml.test.integration.functions.tensor.Conv2DBackwardTest.testConv2DBackwardFilterSparse4</strong></a></li><li><a href='https://sparktc.ibmcloud.com/jenkins/job/SystemML-PullRequestBuilder/2034/org.apache.systemml$systemml/testReport/org.apache.sysml.test.integration.functions.tensor/Conv2DBackwardTest/testConv2DBackwardFilterSparse5/'><strong>org.apache.sysml.test.integration.functions.tensor.Conv2DBackwardTest.testConv2DBackwardFilterSparse5</strong></a></li><li><a href='https://sparktc.ibmcloud.com/jenkins/job/SystemML-PullRequestBuilder/2034/org.apache.systemml$systemml/testReport/org.apache.sysml.test.integration.functions.tensor/Conv2DBackwardTest/testConv2DBackwardFilterSparse6/'><strong>org.apache.sysml.test.integration.functions.tensor.Conv2DBackwardTest.testConv2DBackwardFilterSparse6</strong></a></li><li><a href='https://sparktc.ibmcloud.com/jenkins/job/SystemML-PullRequestBuilder/2034/org.apache.systemml$systemml/testReport/org.apache.sysml.test.integration.functions.tensor/Conv2DBackwardTest/testConv2DBackwardFilterSparse7/'><strong>org.apache.sysml.test.integration.functions.tensor.Conv2DBackwardTest.testConv2DBackwardFilterSparse7</strong></a></li><li><a href='https://sparktc.ibmcloud.com/jenkins/job/SystemML-PullRequestBuilder/2034/org.apache.systemml$systemml/testReport/org.apache.sysml.test.integration.functions.tensor/Conv2DBackwardTest/testConv2DBackwardFilterSparse8/'><strong>org.apache.sysml.test.integration.functions.tensor.Conv2DBackwardTest.testConv2DBackwardFilterSparse8</strong></a></li><li><a href='https://sparktc.ibmcloud.com/jenkins/job/SystemML-PullRequestBuilder/2034/org.apache.systemml$systemml/testReport/org.apache.sysml.test.integration.functions.tensor/Conv2DBackwardTest/testConv2DBackwardFilterSparse9/'><strong>org.apache.sysml.test.integration.functions.tensor.Conv2DBackwardTest.testConv2DBackwardFilterSparse9</strong></a></li><li><a href='https://sparktc.ibmcloud.com/jenkins/job/SystemML-PullRequestBuilder/2034/org.apache.systemml$systemml/testReport/org.apache.sysml.test.integration.functions.tensor/Conv2DBackwardTest/testConv2DBackwardFilterDense1/'><strong>org.apache.sysml.test.integration.functions.tensor.Conv2DBackwardTest.testConv2DBackwardFilterDense1</strong></a></li><li><a href='https://sparktc.ibmcloud.com/jenkins/job/SystemML-PullRequestBuilder/2034/org.apache.systemml$systemml/testReport/org.apache.sysml.test.integration.functions.tensor/Conv2DBackwardTest/testConv2DBackwardFilterDense2/'><strong>org.apache.sysml.test.integration.functions.tensor.Conv2DBackwardTest.testConv2DBackwardFilterDense2</strong></a></li><li><a href='https://sparktc.ibmcloud.com/jenkins/job/SystemML-PullRequestBuilder/2034/org.apache.systemml$systemml/testReport/org.apache.sysml.test.integration.functions.tensor/Conv2DBackwardTest/testConv2DBackwardFilterDense3/'><strong>org.apache.sysml.test.integration.functions.tensor.Conv2DBackwardTest.testConv2DBackwardFilterDense3</strong></a></li><li><a href='https://sparktc.ibmcloud.com/jenkins/job/SystemML-PullRequestBuilder/2034/org.apache.systemml$systemml/testReport/org.apache.sysml.test.integration.functions.tensor/Conv2DBackwardTest/testConv2DBackwardFilterDense4/'><strong>org.apache.sysml.test.integration.functions.tensor.Conv2DBackwardTest.testConv2DBackwardFilterDense4</strong></a></li><li><a href='https://sparktc.ibmcloud.com/jenkins/job/SystemML-PullRequestBuilder/2034/org.apache.systemml$systemml/testReport/org.apache.sysml.test.integration.functions.tensor/Conv2DBackwardTest/testConv2DBackwardFilterDense5/'><strong>org.apache.sysml.test.integration.functions.tensor.Conv2DBackwardTest.testConv2DBackwardFilterDense5</strong></a></li><li><a href='https://sparktc.ibmcloud.com/jenkins/job/SystemML-PullRequestBuilder/2034/org.apache.systemml$systemml/testReport/org.apache.sysml.test.integration.functions.tensor/Conv2DBackwardTest/testConv2DBackwardFilterSparse10/'><strong>org.apache.sysml.test.integration.functions.tensor.Conv2DBackwardTest.testConv2DBackwardFilterSparse10</strong></a></li><li><a href='https://sparktc.ibmcloud.com/jenkins/job/SystemML-PullRequestBuilder/2034/org.apache.systemml$systemml/testReport/org.apache.sysml.test.integration.functions.tensor/Conv2DBackwardTest/testConv2DBackwardFilterSparse11/'><strong>org.apache.sysml.test.integration.functions.tensor.Conv2DBackwardTest.testConv2DBackwardFilterSparse11</strong></a></li><li><a href='https://sparktc.ibmcloud.com/jenkins/job/SystemML-PullRequestBuilder/2034/org.apache.systemml$systemml/testReport/org.apache.sysml.test.integration.functions.tensor/Conv2DBackwardTest/testConv2DBackwardFilterSparse12/'><strong>org.apache.sysml.test.integration.functions.tensor.Conv2DBackwardTest.testConv2DBackwardFilterSparse12</strong></a></li><li><a href='https://sparktc.ibmcloud.com/jenkins/job/SystemML-PullRequestBuilder/2034/org.apache.systemml$systemml/testReport/org.apache.sysml.test.integration.functions.tensor/Conv2DBackwardTest/testConv2DBackwardFilterSparse13/'><strong>org.apache.sysml.test.integration.functions.tensor.Conv2DBackwardTest.testConv2DBackwardFilterSparse13</strong></a></li><li><a href='https://sparktc.ibmcloud.com/jenkins/job/SystemML-PullRequestBuilder/2034/org.apache.systemml$systemml/testReport/org.apache.sysml.test.integration.functions.tensor/Conv2DBackwardTest/testConv2DBackwardFilterSparse14/'><strong>org.apache.sysml.test.integration.functions.tensor.Conv2DBackwardTest.testConv2DBackwardFilterSparse14</strong></a></li><li><a href='https://sparktc.ibmcloud.com/jenkins/job/SystemML-PullRequestBuilder/2034/org.apache.systemml$systemml/testReport/org.apache.sysml.test.integration.functions.tensor/Conv2DBackwardTest/testConv2DBackwardFilterSparse15/'><strong>org.apache.sysml.test.integration.functions.tensor.Conv2DBackwardTest.testConv2DBackwardFilterSparse15</strong></a></li><li><a href='https://sparktc.ibmcloud.com/jenkins/job/SystemML-PullRequestBuilder/2034/org.apache.systemml$systemml/testReport/org.apache.sysml.test.integration.functions.tensor/Conv2DTest/testConv2DDense1SP/'><strong>org.apache.sysml.test.integration.functions.tensor.Conv2DTest.testConv2DDense1SP</strong></a></li><li><a href='https://sparktc.ibmcloud.com/jenkins/job/SystemML-PullRequestBuilder/2034/org.apache.systemml$systemml/testReport/org.apache.sysml.test.integration.functions.tensor/Conv2DTest/testConv2DDense2SP/'><strong>org.apache.sysml.test.integration.functions.tensor.Conv2DTest.testConv2DDense2SP</strong></a></li><li><a href='https://sparktc.ibmcloud.com/jenkins/job/SystemML-PullRequestBuilder/2034/org.apache.systemml$systemml/testReport/org.apache.sysml.test.integration.functions.tensor/Conv2DTest/testConv2DDense3SP/'><strong>org.apache.sysml.test.integration.functions.tensor.Conv2DTest.testConv2DDense3SP</strong></a></li><li><a href='https://sparktc.ibmcloud.com/jenkins/job/SystemML-PullRequestBuilder/2034/org.apache.systemml$systemml/testReport/org.apache.sysml.test.integration.functions.tensor/Conv2DTest/testConv2DDense4SP/'><strong>org.apache.sysml.test.integration.functions.tensor.Conv2DTest.testConv2DDense4SP</strong></a></li><li><a href='https://sparktc.ibmcloud.com/jenkins/job/SystemML-PullRequestBuilder/2034/org.apache.systemml$systemml/testReport/org.apache.sysml.test.integration.functions.tensor/Conv2DTest/testConv2DDense5SP/'><strong>org.apache.sysml.test.integration.functions.tensor.Conv2DTest.testConv2DDense5SP</strong></a></li><li><a href='https://sparktc.ibmcloud.com/jenkins/job/SystemML-PullRequestBuilder/2034/org.apache.systemml$systemml/testReport/org.apache.sysml.test.integration.functions.tensor/Conv2DTest/testConv2DDense6SP/'><strong>org.apache.sysml.test.integration.functions.tensor.Conv2DTest.testConv2DDense6SP</strong></a></li><li><a href='https://sparktc.ibmcloud.com/jenkins/job/SystemML-PullRequestBuilder/2034/org.apache.systemml$systemml/testReport/org.apache.sysml.test.integration.functions.tensor/Conv2DTest/testConv2DDense7SP/'><strong>org.apache.sysml.test.integration.functions.tensor.Conv2DTest.testConv2DDense7SP</strong></a></li><li><a href='https://sparktc.ibmcloud.com/jenkins/job/SystemML-PullRequestBuilder/2034/org.apache.systemml$systemml/testReport/org.apache.sysml.test.integration.functions.tensor/Conv2DTest/testConv2DDense1/'><strong>org.apache.sysml.test.integration.functions.tensor.Conv2DTest.testConv2DDense1</strong></a></li><li><a href='https://sparktc.ibmcloud.com/jenkins/job/SystemML-PullRequestBuilder/2034/org.apache.systemml$systemml/testReport/org.apache.sysml.test.integration.functions.tensor/Conv2DTest/testConv2DDense2/'><strong>org.apache.sysml.test.integration.functions.tensor.Conv2DTest.testConv2DDense2</strong></a></li><li><a href='https://sparktc.ibmcloud.com/jenkins/job/SystemML-PullRequestBuilder/2034/org.apache.systemml$systemml/testReport/org.apache.sysml.test.integration.functions.tensor/Conv2DTest/testConv2DDense3/'><strong>org.apache.sysml.test.integration.functions.tensor.Conv2DTest.testConv2DDense3</strong></a></li><li><a href='https://sparktc.ibmcloud.com/jenkins/job/SystemML-PullRequestBuilder/2034/org.apache.systemml$systemml/testReport/org.apache.sysml.test.integration.functions.tensor/Conv2DTest/testConv2DDense4/'><strong>org.apache.sysml.test.integration.functions.tensor.Conv2DTest.testConv2DDense4</strong></a></li><li><a href='https://sparktc.ibmcloud.com/jenkins/job/SystemML-PullRequestBuilder/2034/org.apache.systemml$systemml/testReport/org.apache.sysml.test.integration.functions.tensor/Conv2DTest/testConv2DDense5/'><strong>org.apache.sysml.test.integration.functions.tensor.Conv2DTest.testConv2DDense5</strong></a></li><li><a href='https://sparktc.ibmcloud.com/jenkins/job/SystemML-PullRequestBuilder/2034/org.apache.systemml$systemml/testReport/org.apache.sysml.test.integration.functions.tensor/Conv2DTest/testConv2DDense6/'><strong>org.apache.sysml.test.integration.functions.tensor.Conv2DTest.testConv2DDense6</strong></a></li><li><a href='https://sparktc.ibmcloud.com/jenkins/job/SystemML-PullRequestBuilder/2034/org.apache.systemml$systemml/testReport/org.apache.sysml.test.integration.functions.tensor/Conv2DTest/testConv2DDense7/'><strong>org.apache.sysml.test.integration.functions.tensor.Conv2DTest.testConv2DDense7</strong></a></li><li><a href='https://sparktc.ibmcloud.com/jenkins/job/SystemML-PullRequestBuilder/2034/org.apache.systemml$systemml/testReport/org.apache.sysml.test.integration.functions.tensor/Conv2DTest/testConv2DSparse1SP/'><strong>org.apache.sysml.test.integration.functions.tensor.Conv2DTest.testConv2DSparse1SP</strong></a></li><li><a href='https://sparktc.ibmcloud.com/jenkins/job/SystemML-PullRequestBuilder/2034/org.apache.systemml$systemml/testReport/org.apache.sysml.test.integration.functions.tensor/Conv2DTest/testConv2DSparse2SP/'><strong>org.apache.sysml.test.integration.functions.tensor.Conv2DTest.testConv2DSparse2SP</strong></a></li><li><a href='https://sparktc.ibmcloud.com/jenkins/job/SystemML-PullRequestBuilder/2034/org.apache.systemml$systemml/testReport/org.apache.sysml.test.integration.functions.tensor/Conv2DTest/testConv2DSparse3SP/'><strong>org.apache.sysml.test.integration.functions.tensor.Conv2DTest.testConv2DSparse3SP</strong></a></li><li><a href='https://sparktc.ibmcloud.com/jenkins/job/SystemML-PullRequestBuilder/2034/org.apache.systemml$systemml/testReport/org.apache.sysml.test.integration.functions.tensor/Conv2DTest/testConv2DSparse1/'><strong>org.apache.sysml.test.integration.functions.tensor.Conv2DTest.testConv2DSparse1</strong></a></li><li><a href='https://sparktc.ibmcloud.com/jenkins/job/SystemML-PullRequestBuilder/2034/org.apache.systemml$systemml/testReport/org.apache.sysml.test.integration.functions.tensor/Conv2DTest/testConv2DSparse2/'><strong>org.apache.sysml.test.integration.functions.tensor.Conv2DTest.testConv2DSparse2</strong></a></li><li><a href='https://sparktc.ibmcloud.com/jenkins/job/SystemML-PullRequestBuilder/2034/org.apache.systemml$systemml/testReport/org.apache.sysml.test.integration.functions.tensor/PoolBackwardTest/testMaxPool2DBackwardSparse1/'><strong>org.apache.sysml.test.integration.functions.tensor.PoolBackwardTest.testMaxPool2DBackwardSparse1</strong></a></li><li><a href='https://sparktc.ibmcloud.com/jenkins/job/SystemML-PullRequestBuilder/2034/org.apache.systemml$systemml/testReport/org.apache.sysml.test.integration.functions.tensor/PoolBackwardTest/testMaxPool2DBackwardSparse2/'><strong>org.apache.sysml.test.integration.functions.tensor.PoolBackwardTest.testMaxPool2DBackwardSparse2</strong></a></li><li><a href='https://sparktc.ibmcloud.com/jenkins/job/SystemML-PullRequestBuilder/2034/org.apache.systemml$systemml/testReport/org.apache.sysml.test.integration.functions.tensor/PoolBackwardTest/testMaxPool2DBackwardSparse3/'><strong>org.apache.sysml.test.integration.functions.tensor.PoolBackwardTest.testMaxPool2DBackwardSparse3</strong></a></li><li><a href='https://sparktc.ibmcloud.com/jenkins/job/SystemML-PullRequestBuilder/2034/org.apache.systemml$systemml/testReport/org.apache.sysml.test.integration.functions.tensor/PoolBackwardTest/testMaxPool2DBackwardSparse4/'><strong>org.apache.sysml.test.integration.functions.tensor.PoolBackwardTest.testMaxPool2DBackwardSparse4</strong></a></li><li><a href='https://sparktc.ibmcloud.com/jenkins/job/SystemML-PullRequestBuilder/2034/org.apache.systemml$systemml/testReport/org.apache.sysml.test.integration.functions.tensor/PoolBackwardTest/testMaxPool2DBackwardSparse5/'><strong>org.apache.sysml.test.integration.functions.tensor.PoolBackwardTest.testMaxPool2DBackwardSparse5</strong></a></li><li><a href='https://sparktc.ibmcloud.com/jenkins/job/SystemML-PullRequestBuilder/2034/org.apache.systemml$systemml/testReport/org.apache.sysml.test.integration.functions.tensor/PoolBackwardTest/testMaxPool2DBackwardSparse6/'><strong>org.apache.sysml.test.integration.functions.tensor.PoolBackwardTest.testMaxPool2DBackwardSparse6</strong></a></li><li><a href='https://sparktc.ibmcloud.com/jenkins/job/SystemML-PullRequestBuilder/2034/org.apache.systemml$systemml/testReport/org.apache.sysml.test.integration.functions.tensor/PoolBackwardTest/testMaxPool2DBackwardSparse7/'><strong>org.apache.sysml.test.integration.functions.tensor.PoolBackwardTest.testMaxPool2DBackwardSparse7</strong></a></li><li><a href='https://sparktc.ibmcloud.com/jenkins/job/SystemML-PullRequestBuilder/2034/org.apache.systemml$systemml/testReport/org.apache.sysml.test.integration.functions.tensor/PoolBackwardTest/testMaxPool2DBackwardSparse8/'><strong>org.apache.sysml.test.integration.functions.tensor.PoolBackwardTest.testMaxPool2DBackwardSparse8</strong></a></li><li><a href='https://sparktc.ibmcloud.com/jenkins/job/SystemML-PullRequestBuilder/2034/org.apache.systemml$systemml/testReport/org.apache.sysml.test.integration.functions.tensor/PoolBackwardTest/testMaxPool2DBackwardSparse9/'><strong>org.apache.sysml.test.integration.functions.tensor.PoolBackwardTest.testMaxPool2DBackwardSparse9</strong></a></li><li><a href='https://sparktc.ibmcloud.com/jenkins/job/SystemML-PullRequestBuilder/2034/org.apache.systemml$systemml/testReport/org.apache.sysml.test.integration.functions.tensor/PoolBackwardTest/testMaxPool2DBackwardDense1/'><strong>org.apache.sysml.test.integration.functions.tensor.PoolBackwardTest.testMaxPool2DBackwardDense1</strong></a></li><li><a href='https://sparktc.ibmcloud.com/jenkins/job/SystemML-PullRequestBuilder/2034/org.apache.systemml$systemml/testReport/org.apache.sysml.test.integration.functions.tensor/PoolBackwardTest/testMaxPool2DBackwardDense2/'><strong>org.apache.sysml.test.integration.functions.tensor.PoolBackwardTest.testMaxPool2DBackwardDense2</strong></a></li><li><a href='https://sparktc.ibmcloud.com/jenkins/job/SystemML-PullRequestBuilder/2034/org.apache.systemml$systemml/testReport/org.apache.sysml.test.integration.functions.tensor/PoolBackwardTest/testMaxPool2DBackwardDense3/'><strong>org.apache.sysml.test.integration.functions.tensor.PoolBackwardTest.testMaxPool2DBackwardDense3</strong></a></li><li><a href='https://sparktc.ibmcloud.com/jenkins/job/SystemML-PullRequestBuilder/2034/org.apache.systemml$systemml/testReport/org.apache.sysml.test.integration.functions.tensor/ReluBackwardTest/testReluBackwardDense1/'><strong>org.apache.sysml.test.integration.functions.tensor.ReluBackwardTest.testReluBackwardDense1</strong></a></li><li><a href='https://sparktc.ibmcloud.com/jenkins/job/SystemML-PullRequestBuilder/2034/org.apache.systemml$systemml/testReport/org.apache.sysml.test.integration.functions.tensor/ReluBackwardTest/testReluBackwardDense2/'><strong>org.apache.sysml.test.integration.functions.tensor.ReluBackwardTest.testReluBackwardDense2</strong></a></li><li><a href='https://sparktc.ibmcloud.com/jenkins/job/SystemML-PullRequestBuilder/2034/org.apache.systemml$systemml/testReport/org.apache.sysml.test.integration.functions.tensor/ReluBackwardTest/testReluBackwardDense3/'><strong>org.apache.sysml.test.integration.functions.tensor.ReluBackwardTest.testReluBackwardDense3</strong></a></li><li><a href='https://sparktc.ibmcloud.com/jenkins/job/SystemML-PullRequestBuilder/2034/org.apache.systemml$systemml/testReport/org.apache.sysml.test.integration.scripts.nn/NNTest/testNNLibrary/'><strong>org.apache.sysml.test.integration.scripts.nn.NNTest.testNNLibrary</strong></a></li></ul>

**Comment by akchinSTC** [comment]
Refer to this link for build results (access rights to CI server needed): 
https://sparktc.ibmcloud.com/jenkins/job/SystemML-PullRequestBuilder/2035/<h2>Failed Tests: <span class='status-failure'>3</span></h2><h3><a name='SystemML-PullRequestBuilder/org.apache.systemml:systemml' /><a href='https://sparktc.ibmcloud.com/jenkins/job/SystemML-PullRequestBuilder/2035/org.apache.systemml$systemml/testReport'>SystemML-PullRequestBuilder/org.apache.systemml:systemml</a>: <span class='status-failure'>3</span></h3><ul><li><a href='https://sparktc.ibmcloud.com/jenkins/job/SystemML-PullRequestBuilder/2035/org.apache.systemml$systemml/testReport/org.apache.sysml.test.integration.functions.misc/IPAConstantFoldingScalarVariablePropagationTest/testConstantFoldingScalarPropagation2IPASecondChance/'><strong>org.apache.sysml.test.integration.functions.misc.IPAConstantFoldingScalarVariablePropagationTest.testConstantFoldingScalarPropagation2IPASecondChance</strong></a></li><li><a href='https://sparktc.ibmcloud.com/jenkins/job/SystemML-PullRequestBuilder/2035/org.apache.systemml$systemml/testReport/org.apache.sysml.test.integration.functions.misc/IPAConstantFoldingScalarVariablePropagationTest/testConstantFoldingScalarPropagation2NoIPASecondChance/'><strong>org.apache.sysml.test.integration.functions.misc.IPAConstantFoldingScalarVariablePropagationTest.testConstantFoldingScalarPropagation2NoIPASecondChance</strong></a></li><li><a href='https://sparktc.ibmcloud.com/jenkins/job/SystemML-PullRequestBuilder/2035/org.apache.systemml$systemml/testReport/org.apache.sysml.test.integration.scripts.nn/NNTest/testNNLibrary/'><strong>org.apache.sysml.test.integration.scripts.nn.NNTest.testNNLibrary</strong></a></li></ul>

**Comment by akchinSTC** [comment]
Refer to this link for build results (access rights to CI server needed): 
https://sparktc.ibmcloud.com/jenkins/job/SystemML-PullRequestBuilder/2060/

**Comment by niketanpansare** [comment]
@nakul02 can you please review this PR ?

**Comment by akchinSTC** [comment]
Refer to this link for build results (access rights to CI server needed): 
https://sparktc.ibmcloud.com/jenkins/job/SystemML-PullRequestBuilder/2062/

File: src/main/cpp/kernels/SystemML.cu [hunk 1]
```diff
@@ -156,6 +156,18 @@ __global__ void relu_backward(double* X,  double* dout, double* ret, int rlen, i
 	}
 }
 
+// Performs inplace addition: ret += input
```
Comment (nakul02) [inline src/main/cpp/kernels/SystemML.cu]: Could you pleae add doxygen style doc here?

**Comment by nakul02** [comment]
Other than that little documentation, LGTM :+1:

**Comment by akchinSTC** [comment]
Refer to this link for build results (access rights to CI server needed): 
https://sparktc.ibmcloud.com/jenkins/job/SystemML-PullRequestBuilder/2067/

**Pull request closed without merging**